cps/kni.c

```diff
@@ -485,3 +486,314 @@ rm_kni(void)
 		RTE_LOG(ERR, GATEKEEPER, "Error removing %s: %s\n",
 			name, strerror(errno));
 }
+
+/*
+ * Processing ARP and ND packets from the KNI. If responding to
+ * an ARP/ND packet fails, we remove the request from the linked list,
+ * forcing the KNI to issue another resolution request.
+ */
+
+static void
+cps_arp_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused)) enum lls_reply_ty ty, int *pcall_again)
+{
+	struct gatekeeper_if *iface = arg;
+	struct rte_mbuf *created_pkt;
+	struct ether_hdr *eth_hdr;
+	struct arp_hdr *arp_hdr;
+	size_t pkt_size;
+	struct cps_config *cps_conf = get_cps_conf();
+	struct rte_kni *kni;
+	struct rte_mempool *mp;
+	struct cps_request *req;
+	int ret;
+
+	if (pcall_again != NULL)
+		*pcall_again = false;
+
+	mp = cps_conf->net->gatekeeper_pktmbuf_pool[
+		rte_lcore_to_socket_id(cps_conf->lcore_id)];
+	created_pkt = rte_pktmbuf_alloc(mp);
+	if (created_pkt == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"cps: could not allocate an ARP reply on the %s KNI\n",
+			iface->name);
+		goto remove;
+	}
+
+	pkt_size = sizeof(struct ether_hdr) + sizeof(struct arp_hdr);
+	created_pkt->data_len = pkt_size;
+	created_pkt->pkt_len = pkt_size;
+
+	/* Set-up Ethernet header. */
+	eth_hdr = rte_pktmbuf_mtod(created_pkt, struct ether_hdr *);
+	ether_addr_copy(&map->ha, &eth_hdr->s_addr);
+	ether_addr_copy(&iface->eth_addr, &eth_hdr->d_addr);
```

Comment:
OK -- just let me know whether you're still reviewing, or if this is the last outstanding comment.